File: LC/[0312] Burst Balloons.java
class Solution {
    public int maxCoins(int[] iNums) {
      int[] nums = new int[iNums.length + 2]; //java new array; java int array
      int n = 1;
      for (int x : iNums) if (x > 0) nums[n++] = x; //to insert into java array -> clone
      nums[0] = nums[n++] = 1;
  
      int[][] dp = new int[n][n];//java two dimentional array
      for (int k = 2; k < n; ++k)
          for (int left = 0; left < n - k; ++left) {
              int right = left + k;
              for (int i = left + 1; i < right; ++i)
                  dp[left][right] = Math.max(dp[left][right], 
                  nums[left] * nums[i] * nums[right] + dp[left][i] + dp[i][right]);//java max
          }
  
          return dp[0][n - 1];
      }
  }

File: LC/[0312] Burst Balloons.ts
function maxCoins(nums: number[]): number {
    const n = nums.length;

    //ts: how to insert to array
    const extendedNums = [1, ...nums, 1]; // Add 1 before and after nums array, ts spreading. ts insert array

    // Initialize the DP table
    //ts: how to initialize an array with defined size
    //ts: how to generate an array
    //ts: generate a list of undefined with size (n+2)
    //ts: each undefined then is mapped to an array of size (n+2) and initialized with 0
    const dp: number[][] = Array.from({ length: n + 2 }, () => Array(n + 2).fill(0));

    // Fill the DP table
    for (let length = 1; length <= n; length++) {
        for (let left = 1; left <= n - length + 1; left++) {
            const right = left + length - 1;
            for (let i = left; i <= right; i++) {
                dp[left][right] = Math.max(
                    dp[left][right],
                    dp[left][i - 1] + extendedNums[left - 1] * extendedNums[i] * extendedNums[right + 1] + dp[i + 1][right]
                );
            }
        }
    }

    return dp[1][n];
}